tee-worker/core/parentchain/indirect-calls-executor/src/executor/litentry/batch_all.rs

```diff
@@ -0,0 +1,422 @@
+// (at your option) any later version.
+//
+// Litentry is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with Litentry.  If not, see <https://www.gnu.org/licenses/>.
+
+// use super::{create_identity::CreateIdentity, set_user_shielding_key::SetUserShieldingKey};
+use crate::{
+	error::{Error, Result},
+	executor::{
+		litentry::{
+			create_identity::CreateIdentity, remove_identity::RemoveIdentity,
+			set_user_shielding_key::SetUserShieldingKey, verify_identity::VerifyIdentity,
+		},
+		Executor,
+	},
+	IndirectCallsExecutor,
+};
+use codec::{Decode, Input};
+use itp_node_api::{
+	api_client::ParentchainUncheckedExtrinsic,
+	metadata::{
+		pallet_imp::IMPCallIndexes,
+		pallet_teerex::TeerexCallIndexes,
+		pallet_utility::UTILCallIndexes,
+		pallet_vcmp::VCMPCallIndexes,
+		provider::{AccessNodeMetadata, Error as metadataProviderError},
+	},
+};
+use itp_sgx_crypto::{key_repository::AccessKey, ShieldingCryptoDecrypt, ShieldingCryptoEncrypt};
+use itp_stf_executor::traits::StfEnclaveSigning;
+use itp_top_pool_author::traits::AuthorApi;
+use itp_types::{
+	extrinsics::ParentchainUncheckedExtrinsicWithStatus, BatchAllFn, CallIndex,
+	CreateIdentityParameters, RemoveIdentityParameters, SetUserShieldingKeyParameters,
+	SupportedCall, VerifyIdentityParameters, H256,
+};
+use litentry_primitives::ParentchainBlockNumber;
+use sp_std::{vec, vec::Vec};
+
+pub(crate) struct BatchAll {
+	pub(crate) block_number: ParentchainBlockNumber,
+}
+
+const V4: u8 = 4;
+
+impl BatchAll {
+	fn execute_internal<
+		ShieldingKeyRepository,
+		StfEnclaveSigner,
+		TopPoolAuthor,
+		NodeMetadataProvider,
+	>(
+		&self,
+		context: &IndirectCallsExecutor<
+			ShieldingKeyRepository,
+			StfEnclaveSigner,
+			TopPoolAuthor,
+			NodeMetadataProvider,
+		>,
+		extrinsic: ParentchainUncheckedExtrinsic<
+			<Self as Executor<
+				ShieldingKeyRepository,
+				StfEnclaveSigner,
+				TopPoolAuthor,
+				NodeMetadataProvider,
+			>>::Call,
+		>,
+	) -> Result<()>
+	where
+		ShieldingKeyRepository: AccessKey,
+		<ShieldingKeyRepository as AccessKey>::KeyType: ShieldingCryptoDecrypt<Error = itp_sgx_crypto::Error>
+			+ ShieldingCryptoEncrypt<Error = itp_sgx_crypto::Error>,
+		StfEnclaveSigner: StfEnclaveSigning,
+		TopPoolAuthor: AuthorApi<H256, H256> + Send + Sync + 'static,
+		NodeMetadataProvider: AccessNodeMetadata,
+		NodeMetadataProvider::MetadataType:
+			IMPCallIndexes + TeerexCallIndexes + VCMPCallIndexes + UTILCallIndexes,
+	{
+		let (_, calls) = extrinsic.function;
+		log::warn!("enter batch call, len:{}", calls.len());
+
+		for call in calls.iter() {
+			match call {
+				itp_types::SupportedCall::SetUserShieldingKey(call_index, Some(params)) => {
```

Comment:
I think this could be further refactored/optimised so that we don't need to do every matched case here, but let's leave it for later

Reply:
I thought about it. But didn't sort it out with a good design to have general 'executor' to check 'is_target_call'. Maybe I need to dig in deeper. 😭